lib/Sema/CSSyntacticElement.cpp

```diff
@@ -227,24 +227,24 @@ class TypeVariableRefFinder : public ASTWalker {
   }
 };
 
-/// Find any references to not yet resolved outer closure parameters
-/// used in the body of the inner closure. This is required because
+/// Find any references to not yet resolved outer VarDecls (including closure
+/// parameters) used in the body of the inner closure. This is required because
 /// isolated conjunctions, just like single-expression closures, have
 /// to be connected to type variables they are going to use, otherwise
 /// they'll get placed in a separate solver component and would never
 /// produce a solution.
-class UnresolvedClosureParameterCollector : public ASTWalker {
+class UnresolvedVarCollector : public ASTWalker {
   ConstraintSystem &CS;
 
   llvm::SmallSetVector<TypeVariableType *, 4> Vars;
 
 public:
-  UnresolvedClosureParameterCollector(ConstraintSystem &cs) : CS(cs) {}
+  UnresolvedVarCollector(ConstraintSystem &cs) : CS(cs) {}
 
   PreWalkResult<Expr *> walkToExprPre(Expr *expr) override {
     if (auto *DRE = dyn_cast<DeclRefExpr>(expr)) {
       auto *decl = DRE->getDecl();
-      if (isa<ParamDecl>(decl)) {
+      if (isa<VarDecl>(decl)) {
```

Comment:
This was actually done on purpose for closures - they only allow parameter inference, everything else either not-yet resolved or unresolved is an error. This shouldn't be a problem in non-closure situation either because conjunctions are resolved only when there are no more type variables/disjunctions to bind. I think this is only comes up when the original result builder transform is enabled because the ordering of inference is different.

Reply:
If you want support that you'd need to guard that on context not being a closure and keep in mind that it would enable bi-direction inferences for variables and could create mismatches between branches?